src/main/java/sjchat/users/UserServiceServer.java

```
@@ -0,0 +1,89 @@
+      @Override
+      public void run() {
+        System.err.println("*** shutting down gRPC server since JVM is shutting down");
+        UserServiceServer.this.stop();
+        System.err.println("*** server shut down");
+      }
+    });
+  }
+
+  private void stop() {
+    if (server != null) {
+      server.shutdown();
+    }
+  }
+
+  /**
+   * Await termination on the main thread since the grpc library uses daemon threads.
+   */
+  private void blockUntilShutdown() throws InterruptedException {
+    if (server != null) {
+      server.awaitTermination();
+    }
+  }
+
+  static class UserService extends UserServiceGrpc.UserServiceImplBase {
+
+    @Override
+    public void createUser(UserDataRequest req, StreamObserver<UserResponse> responseObserver) {
+      Random random = new Random();
+
+      User.Builder userBuilder = User.newBuilder();
+      userBuilder.setId(Math.abs(random.nextInt(100)));
+      userBuilder.setUsername(req.getUsername());
+
+      UserResponse userResponse = UserResponse.newBuilder().setUser(userBuilder).build();
+
+      responseObserver.onNext(userResponse);
+      responseObserver.onCompleted();
+    }
+
+    @Override
+    public void getUser(UserRequest req, StreamObserver<UserResponse> responseObserver) {
+      User.Builder userBuilder = User.newBuilder();
+      userBuilder.setId(req.getId());
+      userBuilder.setUsername("user_" + req.getId());
+
+      UserResponse userResponse = UserResponse.newBuilder().setUser(userBuilder).build();
+
+      responseObserver.onNext(userResponse);
+      responseObserver.onCompleted();
+    }
+
+    @Override
+    public void updateUser(UserDataRequest req, StreamObserver<UserResponse> responseObserver) {
+      User.Builder userBuilder = User.newBuilder();
+      userBuilder.setId(req.getId());
+      userBuilder.setUsername(req.getUsername());
+
+      UserResponse userResponse = UserResponse.newBuilder().setUser(userBuilder).build();
```

If no user exists we can return an error with ``requestObserver.onError(e);``